distributed/span.py

```diff
@@ -0,0 +1,183 @@
+from typing import Callable, Iterator, NamedTuple
+
+_current_span: contextvars.ContextVar[Span] = contextvars.ContextVar("current_span")
+
+
+class Span:
+    label: tuple[str, ...]
+    metric: Callable[[], float]
+    start_time: float | None
+    stop_time: float | None
+    subspans: list[Span]
+    _token: contextvars.Token | None
+
+    def __init__(
+        self,
+        label: str | tuple[str, ...],
+        metric: Callable[[], float] = time.perf_counter,
+    ) -> None:
+        self.label = (label,) if isinstance(label, str) else label
+        self.metric = metric
+        self.start_time = None
+        self.stop_time = None
+        self.subspans = []
+        self._token = None
+
+    @property
+    def total_time(self) -> float:
+        assert self.start_time is not None
+        assert self.stop_time is not None
+        return self.stop_time - self.start_time
+
+    @cached_property
+    def other_time(self) -> float:
+        assert self.start_time is not None
+        assert self.stop_time is not None
+        return sum(span.total_time for span in self.subspans)
+
+    @property
+    def own_time(self) -> float:
+        assert self.start_time is not None
+        assert self.stop_time is not None
+        return self.total_time - self.other_time
+
+    @property
+    def done(self) -> bool:
+        return self.start_time is not None and self.stop_time is not None
+
+    @property
+    def running(self) -> bool:
+        return self.start_time is not None and self.stop_time is None
+
+    def start(self) -> None:
+        assert self.start_time is None
+        assert self.stop_time is None
+        self.start_time = self.metric()
+
+    def stop(self) -> None:
+        assert self.start_time is not None
+        assert self.stop_time is None
+        self.stop_time = self.metric()
+
+    @contextmanager
+    def as_current(self) -> Iterator[None]:
+        self._set_current()
+        try:
+            yield
+        finally:
+            self._unset_current()
+
+    def __enter__(self) -> Span:
+        self.start()
+        self._set_current()
+        return self
+
+    def __exit__(self, *exc: object) -> None:
+        self._unset_current()
+        self.stop()
+
+    def _set_current(self) -> None:
+        assert not self._token
+        self._token = _current_span.set(self)
+
+    def _unset_current(self) -> None:
+        assert self._token
+        _current_span.reset(self._token)
+        self._token = None
+
+    def _subspan(
+        self,
+        label: str | tuple[str, ...],
+        metric: Callable[[], float] = time.perf_counter,
+    ) -> Span:
+        assert (
+            self.start_time is not None
+        ), "Cannot create sub-span for a span that has not started"
+        assert (
+            self.stop_time is None
+        ), "Cannot create sub-span for a span that has already stopped"
+        # TODO allow different metrics, or always use `self.metric`?
```

Comment:
You definitely want to allow different metrics.
e.g.
```python
with meter("thread-wall"):
    with meter("thread-cpu", time.thread_time):
        ...
```

Reply:
Maybe, but thread time is the only time-related case I can think of, and that could be stored in an attribute/metadata instead of a span itself https://github.com/dask/distributed/pull/7619#issuecomment-1458496176